lib/sessions.js

```diff
@@ -243,6 +243,106 @@ class ClientSession extends EventEmitter {
   toBSON() {
     throw new Error('ClientSession cannot be serialized to BSON.');
   }
+
+  /**
+   * A user provided function to be run within a transaction
+   *
+   * @callback WithTransactionCallback
+   * @param {ClientSession} session The parent session of the transaction running the operation. This should be passed into each operation within the lambda.
+   * @returns {Promise} The resulting Promise of operations run within this transaction
+   */
+
+  /**
+   * Runs a provided lambda within a transaction, retrying either the commit operation
+   * or entire transaction as needed (and when the error permits) to better ensure that
+   * the transaction can complete successfully.
+   *
+   * IMPORTANT: This method requires the user to return a Promise, all lambdas that do not
+   * return a Promise will result in undefined behavior.
+   *
+   * @param {WithTransactionCallback} fn
+   * @param {TransactionOptions} [options] Optional settings for the transaction
+   */
+  withTransaction(fn, options) {
+    const self = this;
+    const startTime = Date.now();
+
+    function retryCommit() {
+      return self.commitTransaction().catch(err => {
+        if (
+          !isWriteConcernTimeoutError(err) &&
+          (err instanceof MongoError && err.hasErrorLabel('UnknownTransactionCommitResult')) &&
+          Date.now() - startTime < 120000
```

Comment:
magic # should be a const somewhere, with a comment block explaining why we are going with 2 minutes.

Reply:
Also, maybe instead of doing `Date.now() - startTime < 120000` all over the place, we can instead do `const timeout = Date.now() + 120000`, and then just check `Date.now() >= timeout`

Reply:
120000 is indeed a magic number that will go away in a few months, in favor of a client-side operations timeout. I'm inclined to just leave this as-is at the moment.

Reply:
added `hasNotTimedOut` and `MAX_WITH_TRANSACTION_TIMEOUT`